### PR TITLE
fix get query from body of GET request

### PR DESCRIPTION
**Pull request by tochka**
We found that if use the most popular Golang HTTP driver for Clickhouse (https://github.com/mailru/go-clickhouse) Chproxy doesn't cache requests. It's related to how to get a query for the library the query is passed via the body (for POST and  GET).
https://github.com/mailru/go-clickhouse/blob/e48216604dfdd66cdace05708e5c7c745baf21f7/conn.go#L300

The Clickhouse  HTTP server works correctly if send query into body of a GET request. Also the Clickhouse HTTP server support split query by query URL parameter and body for a GET request.
https://clickhouse.tech/docs/en/interfaces/http/

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -66,26 +68,44 @@ func getQuerySnippet(req *http.Request) string {
...
-	return data
+	if len(query) != 0 && len(data) != 0 {
+		query += "\n"
+	}
+
+	return query + data
 }
 
 // getFullQuery returns full query from req.
 func getFullQuery(req *http.Request) ([]byte, error) {
```
Comment (hagen1778) [inline utils.go:105]: nit:
```
func getFullQuery(req *http.Request) ([]byte, error) {
	var result bytes.Buffer
	if q := req.URL.Query().Get("query"); q != "" {
		result.WriteString(q)
	}
	data, err := ioutil.ReadAll(req.Body)
	if err != nil {
		return nil, err
	}
	// restore body for further reading
	req.Body = ioutil.NopCloser(bytes.NewBuffer(data))
	if u := getDecompressor(req); u != nil {
		br := bytes.NewReader(data)
		b, err := u.decompress(br)
		if err != nil {
			return nil, fmt.Errorf("cannot uncompress query: %s", err)
		}
		data = b
	}
	
	if result.Len() != 0 && len(data) != 0 {
		result.WriteByte('\n')
	}
	result.Write(data)

	return result.Bytes(), nil
}
```

File: utils.go [hunk 2]
```diff
@@ -66,26 +68,44 @@ func getQuerySnippet(req *http.Request) string {
 
 	u := getDecompressor(req)
 	if u == nil {
-		return data
+		if len(query) != 0 && len(data) != 0 {
```
Comment (hagen1778) [inline utils.go:71]: I think these 3 similar checks may be simplified. Consider to split the function into two:
```
func getQuerySnippet(req *http.Request) string {
	query := req.URL.Query().Get("query")
	body, err := getQuerySnippetFromBody(req)
	if len(query) != 0 && len(data) != 0 {
		query += "\n"
	}
	return query + body
```

File: utils.go [hunk 3]
```diff
@@ -47,8 +47,10 @@ func getAuth(req *http.Request) (string, string) {
...
 func getQuerySnippet(req *http.Request) string {
-	if req.Method == http.MethodGet {
-		return req.URL.Query().Get("query")
+	var query string
+
+	if req.URL.Query().Get("query") != "" {
+		query = req.URL.Query().Get("query")
 	}
 
 	crc, ok := req.Body.(*cachedReadCloser)
```
Comment (hagen1778) [inline utils.go:56]: you must check if request is POST or body is not nil. Otherwise process will panic

File: utils_test.go [hunk 4]
```diff
@@ -68,7 +68,7 @@ func testCanCacheQuery(t *testing.T, q string, expected bool) {
 }
 
 func TestGetQuerySnippetGET(t *testing.T) {
-	req, err := http.NewRequest("GET", "", nil)
+	req, err := http.NewRequest("GET", "", bytes.NewBuffer(nil))
```
Comment (hagen1778) [inline utils_test.go:71]: this change shows the issue when `getQuerySnippet` panics if body is nil. Body could be nil for GET requests and proxy will panic in runtime:
```
	// For client requests, a nil body means the request has no
	// body, such as a GET request. 
```

**Review by hagen1778 — CHANGES_REQUESTED** [review]
Pls see review comments

File: utils.go (hunk 2 above)
Comment (tochka) [reply utils.go]: Thank you for your comment. I've fixed

File: utils.go (hunk 1 above)
Comment (tochka) [reply utils.go:94]: I rewrite this code the same like getQuerySnippet from your comment below

File: utils.go (hunk 3 above)
Comment (tochka) [reply utils.go:61]: I revised the source code of HTTP lib. There is a comment that it should never happen. You can check it yourself
https://golang.org/src/net/http/request.go?s=3211:11861#L174

File: utils_test.go (hunk 4 above)
Comment (tochka) [reply utils_test.go:71]: It's true only for the client mode but the server mode is not
https://golang.org/src/net/http/request.go?s=3211:11861#L174

Comment (hagen1778) [reply utils_test.go:71]: True, didn't know about that. However, I still find it useful to check object for not being nil before calling its methods. I think, it makes code more robust. 
Otherwise, you'll need to explain in this test why it is required to path non-nil body for GET request.

**Review by hagen1778 — APPROVED** [review]
lgtm

File: utils_test.go (hunk 4 above)
Comment (tochka) [reply utils_test.go]: Sure, I added check for request body isn't nil and fixed the test
